tests/LinAlg/matrixTestsSparse.hpp

```diff
@@ -422,6 +422,74 @@ class MatrixTestsSparse : public TestBase
     return fail;
   }
 
+ /**
+  * @brief Copy 'n_rows' rows from matrix 'A', started from 'A_rows_st', to the rows started from 'B_rows_st' in 'B'.
+  * The non-zero elements start from 'B_nnz_st' will be replaced by the new elements.
+  *
+  * @pre 'A' must have exactly, or more than 'n_rows' rows after row 'A_rows_st'
+  * @pre 'B' must have exactly, or more than 'n_rows' rows after row 'B_rows_st'
+  * @pre 'B_nnz_st' + the number of non-zeros in the copied the rows must be less or equal to B.nnz
+  * @pre User must know the nonzero pattern of A and B. Assume non-zero patterns of A and B wont change, and A is a submatrix of B
+  * @pre Otherwise, this function may replace the non-zero values and nonzero patterns for the undesired elements.
+  */
+  int
+  copyRowsFromSrcToDest(
+    hiop::hiopMatrixSparse& A,
+    hiop::hiopMatrixSparse& B,
+    local_ordinal_type A_rows_st, local_ordinal_type n_rows,
+    local_ordinal_type B_rows_st, local_ordinal_type B_nnz_st
+    )
+  {
+    const local_ordinal_type* A_iRow = getRowIndices(&A);
+    const local_ordinal_type* A_jCol = getColumnIndices(&A);
+    const local_ordinal_type A_nnz = A.numberOfNonzeros();
+
+    const local_ordinal_type* B_iRow = getRowIndices(&B);
+    const local_ordinal_type* B_jCol = getColumnIndices(&B);
+    const local_ordinal_type B_nnz = B.numberOfNonzeros();
+
+    auto nnz_A_need_to_copy{0};
+    for(auto k=0;k<A_nnz;k++){
+      if(A_iRow[k] >= A_rows_st && A_iRow[k] < A_rows_st + n_rows )
+      {
+        nnz_A_need_to_copy++;
+      }
+      // assume matrix element is ordered by row
+      if(A_iRow[k]>=A_rows_st + n_rows)
+      {
+        break;
+      }
+    }
+
+    assert(A.n() >= B.n());
+    assert(n_rows + A_rows_st <= A.m());
+    assert(n_rows + B_rows_st <= B.m());
+    assert(nnz_A_need_to_copy<A_nnz);
+    assert(nnz_A_need_to_copy+B_nnz_st<=B_nnz);
+
+    const real_type A_val = one;
+    const real_type B_val = half;
+
+    A.setToConstant(A_val);
+    B.setToConstant(B_val);
+
+    auto itnz_src{0};
+    auto itnz_dest=B_nnz_st;
+    int fail{0};
+
+    B.copyRowsFromSrcToDest(A, A_rows_st, n_rows, B_rows_st, B_nnz_st);
+
+    auto val = getMatrixData(&B);
+
+    fail += verifyAnswer(&B,0,B_nnz_st,B_val);
```

Comment:
To be more thorough, we should have another verifyAnswer that takes a lambda for sparse mats, eg
```cpp
fail += verifyAnswer(&B,
  [=] (local_ordinal_type i, local_ordinal_type j) -> real_type
  {
    if(i==0 && j==B_nnz_st) return B_val;
    else if // other two conditions...
    else if // other two conditions...
    else return 0.;
  });
```

Reply:
I am trying to check multiple values in a sparse matrix, not a single value.
The nonzero indices for these nonzero starts from nnz_st to nnz_ed.
I can change it to a lambda function like you mentioned, but I need to assign it to a different name, or use different declaration.
This is because the current implementation assumes the input matrix is dense 
https://github.com/LLNL/hiop/blob/fcd29a04d9b28f57f3ab85f95f969924843a5546/tests/LinAlg/matrixTestsSparseTriplet.cpp#L211-L224

Reply:
I'll submit a pr against your branch to fix this

Reply:
sure. I tried to not touch your existing code and only add my code into the repository.
I believer it will be really helpful if we can fix this issue.
About verifying the answer for a sparse matrix, I do have some different idea. Currently we only verify the nonzero values given by the nonzero index, but I think we may also want to verify the nonzero pattern, For example, verify the kth nonzero is located on some particular row and col.  This may be not urgent, since now we assume the sparsity pattern won't change. If we use sparse matrices correctly, we can skip check the pattern. However, if there is an error in the sparsity pattern when copy to/from another matrix/vector, it will be hard to detect. Please let me know your opinion about this. Cheers,

Reply:
The way I see it there is little use for a `getElement` method for a sparse matrix simply because such method is cumbersome to implement and use. Method `verifyAnswer` is always implementation specific, so it should access matrix data directly, no need to use `getElement` or similar.

Reply:
Verifying sparsity pattern should not be difficult either. Again, I would implement that by accessing matrix data directly and not try to implement `getElement` method for that. Method `getElement` is design to retrieve and check only one single element in a vector or matrix. If you are checking multiple elements, you should be accessing matrix data directly. This is also true for dense matrices.